addons/account/models/account_move_line.py

```diff
@@ -1847,94 +1840,483 @@ def get_accounting_rate(vals):
+                # there is no journal item left with a zero amount residual in one currency but not in the other.
+                # However, after a migration coming from an older version with an older partial reconciliation or due to
+                # some rounding issues (when dealing with different decimal places for example), we could need an extra
+                # exchange difference journal entry to handle them.
+                exchange_lines_to_fix = self.env['account.move.line']
+                amounts_list = []
+                exchange_max_date = date.min
+                for aml in involved_amls:
+                    if not aml.company_currency_id.is_zero(aml.amount_residual):
+                        exchange_lines_to_fix += aml
+                        amounts_list.append({'amount_residual': aml.amount_residual})
+                    elif not aml.currency_id.is_zero(aml.amount_residual_currency):
+                        exchange_lines_to_fix += aml
+                        amounts_list.append({'amount_residual_currency': aml.amount_residual_currency})
+                    exchange_max_date = max(exchange_max_date, aml.date)
+                exchange_diff_values = exchange_lines_to_fix._prepare_exchange_difference_move_vals(
+                    amounts_list,
+                    company=involved_amls.company_id,
+                    exchange_date=exchange_max_date,
+                )
+
+                # Exchange difference for cash basis entries.
+                caba_lines_to_reconcile = None
+                if is_cash_basis_needed(involved_amls.account_id):
+                    caba_lines_to_reconcile = involved_amls._add_exchange_difference_cash_basis_vals(exchange_diff_values)
+
+                # Prepare the exchange difference.
+                if exchange_diff_values['move_values']['line_ids']:
+                    exchange_diff_full_batch_index.append(fulL_batch_index)
+                    exchange_diff_values_list.append(exchange_diff_values)
+                    full_batch['caba_lines_to_reconcile'] = caba_lines_to_reconcile
+
+        # ==== Create the full exchange journal entries ====
+        exchange_moves = self._create_exchange_difference_moves(exchange_diff_values_list)
+        for fulL_batch_index, exchange_move in zip(exchange_diff_full_batch_index, exchange_moves):
+            full_batch = full_batches[full_batch_index]
+            amls = full_batch['amls']
+            full_batch['exchange_move'] = exchange_move
+            exchange_move_lines = exchange_move.line_ids.filtered(lambda line: line.account_id == amls.account_id)
+            full_batch['amls'] |= exchange_move_lines
+
+        # ==== Create the full reconcile ====
+        # Note we are using Command.link and not Command.set because Command.set is triggering an unlink that is
+        # slowing down the assignation of the co-fields. Indeed, unlink is forcing a flush.
```

Comment:
I don't get why we have to say it.
I get that we use `set` in the xmls for data: we are replacing stuff in case of update but here, we are about to create a new `full_reconcile_id`, isn't it supposed to be clean of any other link?


Also, if it's standard to use `set`, how much does it impact the performance in this?
You reduced the time it takes from 30s to 3s (or something like that), for how much % loss is it responsible for?

Reply:
@Levizar Command.set == Command.clear + List(Command.link)
Command.clear is currently making an additional query for each record to retrieve the records to unlink.
We are creating a new record so there is no previous records to unlink. The orm is cleaning that part.
For 1000 records, this speedup the reco x2.5~3